connector/src/rest/connector/libs/tmos/implementation.py

```diff
@@ -0,0 +1,444 @@
+                        port: 22
+                        protocol: ssh
+                    rest:
+                        # specify the rest connector class
+                        class: rest.connector.Rest
+                        ip: 1.2.3.4
+                        credentials:
+                            rest:
+                                username: user
+                                password: password
+
+
+
+    Example
+
+    -------
+
+
+
+        >>> from pyats.topology import loader
+
+        >>> testbed = loader.load('/path/to/testbed.yaml')
+
+        >>> device = testbed.devices['bigip1']
+
+        >>> device.connect(alias='rest', via='rest')
+
+        >>> device.rest.connected
+
+        True
+
+        >>> api_url = '/mgmt/tm/ltm/node'
+
+        >>> nodes = device.rest.get(api_url)
+
+        >>> nodes.status_code
+
+        >>> nodes.json()
+    """
+
+    def __init__(self, *args, **kwargs):
+
+        """__init__ instantiates a single connection instance."""
+
+        # instanciate BaseConnection
+
+        # (could use super...)
+
+        BaseConnection.__init__(self, *args, **kwargs)
+
+        self._is_connected = False
+
+    @property
+    def connected(self):
+
+        """Is a device connected"""
+
+        return self._is_connected
+
+    def connect(self, *args, **kwargs):
+
+        """connect to the device via REST"""
+
+        username, password = get_username_password(self)
+        icr_session = iControlRESTSession(username, password)
+        payload = {"command": "run", "utilCmdArgs": '-c "runlevel"'}
+        ip = self.connection_info["ip"].exploded
+        port = self.connection_info.get("port", "443")
+        self.base_url = "https://{ip}:{port}/".format(ip=ip, port=port)
+        login_url = "{f}mgmt/tm/util/bash".format(f=self.base_url)
+
+        log.info(
+            "Connecting to '{d}' with alias "
+            "'{a}'".format(d=self.device.name, a=self.alias)
+        )
+
+        # Connect to the device
+        response = icr_session.post(login_url, json=payload)
+        log.info(response)
+
+        # Make sure it returned ok
+        if not response.ok:
+            # Something bad happened
+            raise iControlUnexpectedHTTPError(
+                "Connection to '{ip}' has returned the "
+                "following code '{c}', instead of the "
+                "expected status code 'ok'".format(
+                    ip=ip, c=response.status_code
+                )
+            )
+
+        self._is_connected = True
+        log.info("Connected successfully to '{d}'".format(d=self.device.name))
+
+        return self._is_connected
+
+    def disconnect(self):
+
+        """disconnect the device for this particular alias"""
+
+        raise NotImplementedError
+
+    def get(self, api_url, timeout=30, verbose=False):
+
+        """GET REST Command to retrieve information from the device"""
+
+        username, password = get_username_password(self)
```

Comment:
Instead of reconnecting every time, you can save self.icr_session in the object and re-use it. 

However if you lose connection to it after x amount of time, you can still place all this code in connect, and just call connect before doing your get action.

This way there will be no code repetition. Right now there are many lines which are repeated in a few methods.

Same comment about the other methods you have.